collector/src/main/java/com/bitmovin/analytics/retryBackend/RetryBackend.kt

```diff
@@ -0,0 +1,87 @@
+package com.bitmovin.analytics.retryBackend
+
+import android.os.Handler
+import android.os.SystemClock
+import android.util.Log
+import com.bitmovin.analytics.data.AdEventData
+import com.bitmovin.analytics.data.Backend
+import com.bitmovin.analytics.data.CallbackBackend
+import com.bitmovin.analytics.data.EventData
+import java.net.ConnectException
+import java.net.SocketTimeoutException
+import java.net.UnknownHostException
+import java.util.Date
+import kotlin.Exception
+import okhttp3.internal.http2.StreamResetException
+
+class RetryBackend(private val next: CallbackBackend, val handler: Handler) : Backend {
+
+    private val TAG = "RetryBackend"
+    private var retryDateToken: Date? = null
+    private val retryQueue = RetryQueue()
+    fun getNextScheduledTime() = retryQueue.getNextScheduleTime()
+
+    override fun send(eventData: EventData) {
+        scheduleSample(RetrySample(eventData, 0, Date(), 0))
+    }
+
+    override fun sendAd(eventData: AdEventData) {
+        scheduleSample(RetrySample(eventData, 0, Date(), 0))
+    }
+
+    private fun scheduleSample(retrySample: RetrySample<Any>) {
+        val callback = object : OnFailureCallback {
+            override fun onFailure(e: Exception, cancel: () -> Unit) {
+                if (e is SocketTimeoutException || e is ConnectException || e is StreamResetException || e is UnknownHostException) {
+                    cancel()
+                    retryQueue.addSample(retrySample)
+                    processQueuedSamples()
+                }
+            }
+        }
+
+        if (retrySample.eventData is EventData) {
+            Log.d(TAG, "sending sample ${retrySample.eventData?.sequenceNumber} retry ${retrySample.retry}")
+            retrySample.eventData.retryCount = retrySample.retry
+            this.next.send(retrySample.eventData, callback)
+        } else if (retrySample.eventData is AdEventData) {
+            retrySample.eventData.retryCount = retrySample.retry
+            this.next.sendAd(retrySample.eventData, callback)
+        }
+    }
+
+    @Synchronized
+    fun processQueuedSamples() {
+        try {
+            val nextScheduledTime = getNextScheduledTime()
+            if (nextScheduledTime != null || retryDateToken != nextScheduledTime) {
+                if (retryDateToken != null) {
+                    handler.removeCallbacks(processSampleRunnable, retryDateToken)
+                }
+                retryDateToken = nextScheduledTime
+                val delay = maxOf((nextScheduledTime?.time ?: 0) - Date().time, 0) // to prevent negative delay
+                handler.postAtTime(processSampleRunnable, retryDateToken, SystemClock.uptimeMillis() + delay)
+            }
+        } catch (e: Exception) {
+            Log.d(TAG, "processQueuedSamples ${e.message}")
```

Comment:
better error message would be useful here (at least that an exception happened).
Also i think you can pass the exception so you get a stacktrace etc. This should not be `Log.d` but `Log.e`
https://developer.android.com/reference/android/util/Log#e(java.lang.String,%20java.lang.String,%20java.lang.Throwable)

Reply:
changed logs in https://github.com/bitmovin/bitmovin-analytics-collector-android/pull/132/commits/479546852b7e47cf387dec0b6ff9b8b7f909c831